src/NServiceBus.Core/Serializers/XML/Deserializer.cs

```diff
@@ -0,0 +1,638 @@
+﻿namespace NServiceBus.Serializers.XML
+{
+    using System;
+    using System.Collections;
+    using System.Collections.Generic;
+    using System.IO;
+    using System.Linq;
+    using System.Reflection;
+    using System.Runtime.Serialization;
+    using System.Xml;
+    using System.Xml.Linq;
+    using NServiceBus.Logging;
+    using NServiceBus.MessageInterfaces;
+    using NServiceBus.Utils.Reflection;
+
+    class Deserializer
+    {
+        public Deserializer(IMessageMapper mapper, XmlSerializerCache cache)
+        {
+            this.mapper = mapper;
+            this.cache = cache;
+        }
+
+        public bool SkipWrappingRawXml { get; set; }
+
+        public bool SanitizeInput { get; set; }
+
+        public object[] Deserialize(Stream stream, IList<Type> messageTypesToDeserialize = null)
+        {
+            if (stream == null)
+            {
+                return null;
+            }
+
+            var result = new List<object>();
+
+            var doc = new XmlDocument
+            {
+                PreserveWhitespace = true
+            };
+
+            var reader = SanitizeInput
+                ? XmlReader.Create(new XmlSanitizingStream(stream), new XmlReaderSettings
+                {
+                    CheckCharacters = false
+                })
+                : XmlReader.Create(stream, new XmlReaderSettings
+                {
+                    CheckCharacters = false
+                });
+
+            doc.Load(reader);
+
+            if (doc.DocumentElement == null)
+            {
+                return result.ToArray();
+            }
+
+            foreach (XmlAttribute attr in doc.DocumentElement.Attributes)
+            {
+                if (attr.Name == "xmlns")
+                {
+                    defaultNameSpace = attr.Value.Substring(attr.Value.LastIndexOf("/") + 1);
+                }
+                else
+                {
+                    if (attr.Name.Contains("xmlns:"))
+                    {
+                        var colonIndex = attr.Name.LastIndexOf(":");
+                        var prefix = attr.Name.Substring(colonIndex + 1);
+
+                        if (prefix.Contains(BASETYPE))
+                        {
+                            var baseType = mapper.GetMappedTypeFor(attr.Value);
+                            if (baseType != null)
+                            {
+                                messageBaseTypes.Add(baseType);
+                            }
+                        }
+                        else
+                        {
+                            prefixesToNamespaces[prefix] = attr.Value;
+                        }
+                    }
+                }
+            }
+
+            if (doc.DocumentElement.Name.ToLower() != "messages")
```

Comment:
Maybe leave high level if/else here but introduce descriptive method for condition and processing bodies of this if/else branch to line 138